src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/EServicesApiServiceImpl.scala

```diff
@@ -0,0 +1,132 @@
+    implicit
+    contexts: Seq[(String, String)],
+    toEntityMarshallerProblem: ToEntityMarshaller[Problem],
+    toEntityMarshallerCatalogEServices: ToEntityMarshaller[CatalogEServices]
+  ): Route = {
+    val result = for {
+      requesterId       <- getOrganizationIdFutureUUID(contexts)
+      apiStates         <- parseArrayParameters(states).traverse(EServiceDescriptorState.fromValue).toFuture
+      pagedResults      <- catalogProcessService.getEServices(
+        name = q,
+        producersIds = parseArrayParameters(producersIds),
+        states = apiStates.map(CatalogProcess.EServiceDescriptorState.fromApi),
+        offset = offset,
+        limit = limit
+      )
+      enhancedEServices <- Future.traverse(pagedResults.eservices)(enhanceEService(requesterId))
+    } yield CatalogEServices(
+      eservices = enhancedEServices,
+      pagination = Pagination(offset = offset, limit = limit, totalResults = pagedResults.totalCount)
+    )
+
+    onComplete(result) {
+      handleError(s"Error retrieving Catalog EServices") orElse { case Success(eServices) =>
+        getEServicesCatalog200(eServices)
+      }
+    }
+  }
+
+  private def enhanceEService(
+    requesterId: UUID
+  )(eService: CatalogProcess.EService)(implicit contexts: Seq[(String, String)]): Future[CatalogEService] = for {
+    // TODO Use directly the tenant once the name will be added to its model
+    producerTenant      <- tenantManagementService.getTenant(eService.producerId)
+    selfcareId          <- producerTenant.selfcareId.toFuture(MissingSelfcareId(producerTenant.id))
+    producerInstitution <- partyProcessService.getInstitution(selfcareId)
+    // End TODO
+
+    requesterTenant <-
+      if (requesterId != eService.producerId) tenantManagementService.getTenant(requesterId)
+      else Future.successful(producerTenant)
+
+    activeDescriptor = eService.descriptors
+      .filter(d => ACTIVE_DESCRIPTOR_STATES_FILTER.contains(d.state))
+      .sortBy(_.version.toInt)
+      .lastOption
+
+    agreement <- activeDescriptor.flatTraverse(d =>
+      agreementProcessService
+        .getAgreements(
+          consumerId = requesterId.toString.some,
+          eServiceId = eService.id.toString.some,
+          descriptorId = d.id.toString.some,
+          states = Nil
+        )
+        .map(_.headOption)
+    )
+  } yield CatalogEService(
+    id = eService.id,
+    name = eService.name,
+    description = eService.description,
+    producer = SlimOrganization(id = eService.producerId, name = producerInstitution.description),
+    agreement = agreement.map(a => SlimAgreement(id = a.id, state = a.state.toApi)),
+    isMine = eService.producerId == requesterId,
+    canSubscribe = certifiedAttributesSatisfied(eService, requesterTenant),
+    activeDescriptor = activeDescriptor.map(d => SlimDescriptor(id = d.id, state = d.state.toApi, d.version))
+  )
+
+  // TODO These methods could be published by the Agreement Process
```

Comment:
Opinions?
If you agree I could do this tomorrow

Reply:
Not clear why moving these to the agreement process, no agreements are involved here.
Probably I'm missing something.

Reply:
Ok, but why not put this in the commons? I mean, I think that adding an endpoint just to check something not strictly related to the domain model is out of the scope of the process.
I agree with you about centralizing it, but I am not sure that I want to call an external service to check it.
These functions are not only used by the agreement process.

Reply:
Nice, two comments at the same time in the exact opposite directions 😂 
I would not put this in commons, this is a business logic strictly related to the agreements, despite being used by other services.
As we centralized persistence models in each Managements, I would in case centralize business logics in the Processes.

But I also had the same doubt pointed out by Antonio, transitive dependencies.
If we keep versions always aligned it should not be a problem, but it already happened that we updated just part of the dependencies, and these transitive deps could be risky.

So we have 3 bad options:
- code duplication
- external business logic module (but transitive deps)
- additional process endpoint (inefficient and out of process scope)

which is the lesser evil?

Reply:
With this amount of code it's `code duplication` for sure